### PR TITLE
2973 log user

### PR DESCRIPTION
**Pull request by iilyak**
Change the log format from `chttpd` to include username.

### CODE REVIEW

File: src/chttpd.erl [hunk 1]
```diff
@@ -1074,3 +1075,53 @@ respond_(#httpd{mochi_req = MochiReq}, Code, Headers, _Args, start_response) ->
     MochiReq:start_response({Code, Headers});
 respond_(#httpd{mochi_req = MochiReq}, Code, Headers, Args, Type) ->
     MochiReq:Type({Code, Headers, Args}).
+
+get_user(#httpd{user_ctx = #user_ctx{name = User}}) -> User;
+get_user(#httpd{user_ctx = undefined}) -> "undefined".
```
Comment (kxepal) [inline src/chttpd.erl]: What's the case for undefined? I thought we always have user_ctx set...

Comment (kxepal) [inline src/chttpd.erl]: Nvm, I see it. Thought it's strange that this value is unstable depending on what operation you proceed.

**Comment by kxepal** [comment]
+1

**Comment by kxepal** [comment]
May be one addition to not log user names as UTF-8, but in some JSON to not get fooled by unicode tricky usernames or get log lines broken by newlines in username.

**Comment by kxepal** [comment]
Oh yes, user names `"null"` and  `"undefined"` will cause a lot of fun.

**Comment by iilyak** [comment]
@kxepal Supporting utf-8 is a very valid concern. Do we actually support it? It doesn't seem possible to transmit utf-8 in a http header. We use basic auth which is based on headers. There is a new [RFC7617](https://datatracker.ietf.org/doc/rfc7617/) which is going to support utf-8. But currently it is not supported. Therefore I do believe that we shouldn't have any utf-8 users in the wild. As utf-8 support is slowly coming maybe we should consider userid validation on couch side to sanitize user's input before it became a problem. The proposed [RFC7613](https://datatracker.ietf.org/doc/rfc7613/?include_text=1) defines what can be in a userid and what shouldn't be there.

**Comment by iilyak** [comment]
@kxepal: I did create a jira ticket [COUCHDB-2974](https://issues.apache.org/jira/browse/COUCHDB-2974) to track utf-8 support.

**Comment by iilyak** [comment]
> @kxepal: Oh yes, user names "null" and "undefined" will cause a lot of fun.

Would you recommend changing `"undefined"` -> `"null"`?

**Comment by kxepal** [comment]
>>Oh yes, user names "null" and "undefined" will cause a lot of fun.

> Would you recommend changing "undefined" -> "null"?

No, I mean that user with name "null" will looks in logs as the anonymous user - so you'll never trace him. Same story with "undefined". I don't think there is any solution except to quote the name or do something to distinguish authed user from anonymous ones.

**Comment by kxepal** [comment]
As for UTF-8: yes, we do. More over, we already support UTF-8 in basic auth header regardless the fact that RFC assumes ASCII there.

**Comment by iilyak** [comment]
@kxepal: Should we log `hex(base64(name))` instead then? Even ASCII could have `\a` or `\r` which could brake the log lines. It such a pity that we historically do not validate user names.

**Comment by kxepal** [comment]
@iilyak I would prefer json or url encoded way. Both are fine to notice tricky bits, but each is useful for further copy-paste into some script to work with it and both provides more-or-less readable close to real variants of username. Plain ASCII ones will remain plain ASCII ones.

However, that could be not very friendly for the end users (admins/devops) to work with such kind of data. Need to think about and call for more opinions around. For instance, what tools ElasticSearch could provide to work with such kind of encoded usernames?

**Comment by rnewson** [comment]
we could apply url encoding before logging it. That way administrators can decode to the real value but we don't have to worry about any unicode trickery.

**Comment by kxepal** [comment]
+1 url encoding then. That's better than json as you can simply copy-paste the name to url and get the doc, it's still readable for ASCII names and quite easy to convert into UTF-8 string if need.

File: src/chttpd.erl [hunk 2]
```diff
@@ -1074,3 +1075,60 @@ respond_(#httpd{mochi_req = MochiReq}, Code, Headers, _Args, start_response) ->
     MochiReq:start_response({Code, Headers});
 respond_(#httpd{mochi_req = MochiReq}, Code, Headers, Args, Type) ->
     MochiReq:Type({Code, Headers, Args}).
+
+get_user(#httpd{user_ctx = #user_ctx{name = User}}) ->
+    mochiweb_util:quote_plus(User);
```
Comment (rnewson) [inline src/chttpd.erl]: couch_util:url_encode ?

**Comment by rnewson** [comment]
+1

File: src/chttpd.erl [hunk 3]
```diff
@@ -1074,3 +1075,60 @@ respond_(#httpd{mochi_req = MochiReq}, Code, Headers, _Args, start_response) ->
...
+    ok = meck:new(couch_log, [passthrough]),
+    ok = meck:expect(couch_log, notice, fun(Format, Args) ->
+        lists:flatten(io_lib:format(Format, Args))
+    end),
+    Message = maybe_log(Req, Resp),
+    ok = meck:unload(couch_log),
+    Message.
+
+
+
```
Comment (kxepal) [inline src/chttpd.erl]: Reserved for future tests? (:

**Comment by kxepal** [comment]
+1, but need to rebase. Bonus +1 if empty space will get reduced (;

**Pull request merged**